tutorials/haskell/haskell-time-tutorial/tutorial.md

```diff
@@ -0,0 +1,537 @@
+>10 to 1. We are constantly reading old code as part of the effort
+>to write new code. ...Therefore, making it easy to read makes it easier
+>to write.”
+
+So this was our motivation, to make something similar to the
+utilities that RoR has and to ease Haskell developer’s lives.
+
+We started by opening an [issue][issue] on GitHub. Quick parenthesis here,
+when you want to work on OSS I’d recommend to start opening an issue
+in the repo and discuss/brainstorm your idea with the maintainers,
+sometimes there are some good reasons why the feature you’re proposing
+have not been yet implemented and if you open a PR directly it could
+rejected, so save yourself a disappointment and open an issue first.
+Back to our issue, the maintainer of the library surprisingly proposed
+to implement these helper functions using pattern synonyms and I said
+surprisingly because I haven’t heard about them, so that’s why I’ll
+like to introduce you some basics about this Haskell feature that,
+actually, it has been around for some years now.
+
+
+## A little about pattern synonyms
+
+First of all I’d like to start with a caveat, this guide is not meant
+to explain pattern synonyms in depth, instead it’s meant to be a practical
+introduction to this concept.
+
+The [GHC documentation][ghc-patterns] describes pattern synonyms like:
+
+> “Pattern synonyms enable giving names to parametrized pattern schemes.
+> They can also be thought of as abstract constructors that don’t have a
+> bearing on data representation.”
+
+And probably we’re in the same place as in the beginning so I’ve preferred
+to introduce pattern synonyms with a real world example. We’re going to
+borrow a pattern that exists in the time library. The [MonthOfYear][month-pattern]
+pattern allows us to define months of the year not by a number but by its name.
+
+The `MonthOfYear` data type is defined like this:
+
+```haskell
+type MonthOfYear = Int
+```
+
+As you can see, it is only an alias for `Int`. We now are able to define
+months like `Int`s where January will correspond to 1 and so on. But
+fortunately, since GHC 7.8 we are able to use pattern synonyms and we
+could, instead of using raw `Int`s, write this down in a pattern style
+and have a more idiomatic way of defining months.
+
+```haskell
+pattern January :: MonthOfYear
+pattern January = 1
+
+pattern February :: MonthOfYear
+pattern February = 2
+
+pattern March :: MonthOfYear
+pattern March = 3
+.
+.
+.
+
+-- | The twelve 'MonthOfYear' patterns form a @COMPLETE@ set.
+pattern December :: MonthOf
+pattern December = 12
+
+{-# COMPLETE January, February, March, April, May, June, July, August, September, October, November, December #-}Year
+```
+
+If you’re curious about why the `{-# COMPLETE #-}` pragma is being
+used, see the [COMPLETE pragma documentation][complete-pragma] for more
+information.
+
+Just to make sure that `November` and `11` are equivalent, let's test it
+in a stack repl.
```

Comment:
```suggestion
in a stack repl:
```

I don't think this is being used for other examples, but I think switching to colon before code snippets should be better.

Reply:
@jpvillaisaza answering since it does not let me to do it directly in your comment. Regarding the laws all the ones you mentioned are required, the ones I added are more like properties that some instances should satisfy. Not really sure how to proceed with this one.

Reply:
Okay, that sounds good.